src/ast/ast_validations.c

```diff
@@ -179,34 +179,169 @@ static AST_Validation _Validate_MATCH_Clause_ReusedEdges(const cypher_astnode_t
+			const cypher_astnode_t *edge = cypher_ast_pattern_path_get_element(path, i);
+			// Make sure relation identifier is unique
+			res = _Validate_MATCH_Clause_ReusedEdges(edge, identifiers, reason);
+			if (res != AST_VALID) break;
+
+			// Validate the range specified by the relation
+			const cypher_astnode_t *range = cypher_ast_rel_pattern_get_varlength(edge);
+			if (range) {
+				res = _MATCH_Clause_Validate_Range(range, reason);
+				if (res != AST_VALID) break;
+			}
+
+			// Validate that the relation is explicit and directed
+			// TODO Lift this limitation when possible
+			if (cypher_ast_rel_pattern_get_direction(edge) == CYPHER_REL_BIDIRECTIONAL) {
+				asprintf(reason, "RedisGraph does not currently support undirected relations.");
+				res = AST_INVALID;
+				break;
+			};
+		}
+	}
+
+	return res;
+}
 
-    // Make sure relation identifier is unique.
-    if(node_type == CYPHER_AST_REL_PATTERN) {
-        if(_Validate_MATCH_Clause_ReusedEdges(node, identifiers, reason) == AST_INVALID) return AST_INVALID;
-    } else if (node_type == CYPHER_AST_RANGE) {
-        return _MATCH_Clause_Validate_Range(node, reason);
-    }
+static AST_Validation _ValidatePattern(const cypher_astnode_t *pattern, TrieMap *identifiers, char **reason) {
+	AST_Validation res = AST_VALID;
+	uint path_count = cypher_ast_pattern_npaths(pattern);
+	for (uint i = 0; i < path_count; i ++) {
+		res = _ValidatePath(cypher_ast_pattern_get_path(pattern, i), identifiers, reason);
+		if (res != AST_VALID) break;
+	}
+    return res;
+}
 
-    // Keep scanning for relation pattern nodes.
-    unsigned int child_count = cypher_astnode_nchildren(node);
-    for(int i = 0; i < child_count; i++) {
-        const cypher_astnode_t *child = cypher_astnode_get_child(node, i);
-        if(_Validate_MATCH_Clause_Relations(child, identifiers, reason) == AST_INVALID) return AST_INVALID;
-    }
+static AST_Validation _ValidateInlinedProperties(const cypher_astnode_t *props, char **reason) {
+	cypher_astnode_type_t type = cypher_astnode_type(props);
+	if (type == CYPHER_AST_PARAMETER) {
+		asprintf(reason, "Parameters are not currently supported in RedisGraph");
+		return AST_INVALID;
+	}
+
+	if (type != CYPHER_AST_MAP) {
+		// This should be impossible
+		asprintf(reason, "Encountered unhandled type in inlined properties.");
+		return AST_INVALID;
+	}
+
+	uint filter_count = cypher_ast_map_nentries(props);
+	for (uint i = 0; i < filter_count; i ++) {
+		cypher_astnode_type_t value_type = cypher_astnode_type(cypher_ast_map_get_value(props, i));
+		if (value_type == CYPHER_AST_IDENTIFIER) {
+			asprintf(reason, "Parameters are not currently supported in RedisGraph");
+			return AST_INVALID;
+		}
+	}
+
+	return AST_VALID;
+}
 
-    return AST_VALID;
+static AST_Validation _ValidateInlinedPropertiesOnPath(const cypher_astnode_t *path, char **reason) {
+	uint path_len = cypher_ast_pattern_path_nelements(path);
+	// Check all entities on the path
+	for (uint i = 0; i < path_len; i ++) {
```

Comment:
might be better to divide this into two for loops, one which will go over nodes and another which will scan edges
this way, we won't be messing with the branch predictor and save computing of modulo